src/compiler/binder.ts

```diff
@@ -3049,7 +3049,11 @@ namespace ts {
         function bindSpecialPropertyAssignment(node: BindablePropertyAssignmentExpression) {
             // Class declarations in Typescript do not allow property declarations
             const parentSymbol = lookupSymbolForPropertyAccess(node.left.expression, container) || lookupSymbolForPropertyAccess(node.left.expression, blockScopeContainer) ;
-            if (!isInJSFile(node) && !isFunctionSymbol(parentSymbol)) {
+            const isFunctionSym = isFunctionSymbol(parentSymbol);
+            if (!isInJSFile(node) && !isFunctionSym) {
+                return;
+            }
+            if (isFunctionSym && isBindableStaticNameExpression(node.left) && getDeclarationName(node.left) === "length") {

```

Comment:
Ideally we would be able to query the type of `Function` ahead of time, but this is done in the binder. Is there some machinery to do this later in the checker? @rbuckton @ahejlsberg

Reply:
We have the ability to introduce new bindings in the checker for properties coming from literal types or `unique symbol` types, but not for this case. 

This seems like an acceptable way to do this, since `length` is a special case per spec (since it is `writable: false, configurable: true`). We should probably also do this for `name`, since it has the same restrictions.

Reply:
Yeah, `name` was what I was thinking of. I just generally don't like these special-cases.

Reply:
I've extended the logic to handle `name` as well.

Reply:
The whole concept of "special property assignments" is a special case. In for a penny...